service: Support `.yaml` extension

cathay4t opened the pull request:
Introduce the support of file named with `.yaml` extension for desired
state file in `/etc/nmstate` folder.

When two files holding the same name but one with `.yml` and another
with `.yaml`, nmstate.service will fail with error message indicating
the confliction. The reason of raising error for this use case is
because both files will be renamed to `foo.applied`. Changing the
renamed file to `foo.yaml.applied` will be change of behavior which is
too risky comparing the corner case of having two YAMLs holding the same
name with different extensions..

Integration test case included.

ihuguet commented:
I don't think that picking one randomly is the best behaviour. It would be better just to use both, they are different files. We don't know if it's an error from the user or not, and even if it's an error, picking one is not a good solution, anyway.

cathay4t commented:
They are different file, but their applied file is the same `foo.applied`. Let me amend the patch to raise error when user has conflict instead of randomly choose.

ihuguet commented:
> They are different file, but their applied file is the same foo.applied. Let me amend the patch to raise error when user has conflict instead of randomly choose.

In that case we could name them foo.yml.applied and foo.yaml.applied.

I'm also OK with raising an error, though. I don't think that users will create foo.yml and foo.yaml at the same time on purpose.

cathay4t commented:
Changing `foo.applied` to `foo.yml.applied` is a change of behavior. So I think raise error is better.

tyll commented:
The commit message is missing a justification for this and without it it also seems to be a bad idea.

cathay4t commented:
Commit message updated with explanation of why we should raise error instead of supporting two files holding the same and with different extensions.

tyll commented:
I meant a justification why it is a good idea to support both .yml and .yaml extensions in the first place.

ihuguet commented:
> I meant a justification why it is a good idea to support both .yml and .yaml extensions in the first place.

Both extensions are valid and common for YAML files. Some people use one, others use the other, and others use sometimes one and sometimes the other. It's quite unconvenient to support only one because people need to remember which one is the valid and which one not. So, the question is the contrary: why was it a good idea to support only one in the first place?

tyll commented:
> Both extensions are valid and common for YAML files. Some people use one, others use the other, and others use sometimes one and sometimes the other. It's quite unconvenient to support only one because people need to remember which one is the valid and which one not. So, the question is the contrary: why was it a good idea to support only one in the first place?

The justification for the initial code should be in the respective commit message. Regarding the invonvenience: Who would need to remember this and when? The directory or README do not provide any indication that they accept arbitrary YAML files, even that they require a specific suffix at all. The usage information in the manpage clearly states that the files require the yml suffix. Once the first file is there, it is very easy to deduct the suffix from the existing file without having to remember anything. Adding unjustified features is bad practice as it leads to unnecessary costs in code complexity, documentation complexity and costs for running unnecessary tests.

cathay4t commented:
We have `man 8 nmstate.service` and `/etc/nmstate/README` explaining the whole work flow. I don't want to create example yaml file there because I think manpage and README file is enough.

cathay4t committed:
service: Support `.yaml` extension

Introduce the support of file named with `.yaml` extension for desired
state file in `/etc/nmstate` folder.

When two files holding the same name but one with `.yml` and another
with `.yaml`, nmstate.service will fail with error message indicating
the confliction. The reason of raising error for this use case is
because both files will be renamed to `foo.applied`. Changing the
renamed file to `foo.yaml.applied` will be change of behavior which is
too risky comparing the corner case of having two YAMLs holding the same
name with different extensions.

Integration test case included.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
The request is coming from OpenShift where customer use `foo.yaml` and complains as not working.

I think it is small issue with clear expectation, I did not request them to create issue for business justification.

Even without business usage, I think it is reasonable to support both extensions.
